base/lttemplates.dtx

```diff
@@ -351,12 +351,41 @@
 %   In the final argument of \cs{DeclareTemplateCode} the assignment of
 %   keys defined by the template may be delayed by including the command
 %   \cs{AssignTemplateKeys}. If this is \emph{not} present, keys are assigned
-%   immediately before the template code. If \cs{AssignTemplateKeys} is
+%   immediately before the template code. If an
+%    \cs{AssignTemplateKeys} command is
 %   present, assignment is delayed until this point. Note that the
 %   command must be \emph{directly} present in the code, not placed
 %   within a nested command/macro.
 % \end{function}
 %
+
+% \begin{function}{\SetTemplateKeys}
+%   \begin{syntax}
+%    \cs{SetTemplateKeys} \Arg{type} \Arg{template} \Arg{keyvals}
+%   \end{syntax}
+
+%    In the final argument of \cs{DeclareTemplateCode} one can
+%    also overwrite the current template key value settings by using the command
+%    \cs{SetTemplateKeys}, i.e., after the template default
+%    values and the values from the instance have been assigned.
+
+%    The main use case for this command is the situation where there
+%    is an argument (normally \texttt{\#1}) to to the template in which
+%    a key/value list can be specified that overwrites the normal
+%    settings. In that case one could use something like
+%\begin{quote}
+%  \verb/\tl_if_empty:nF {#1} { \SetTemplateKeys/\Arg{type}\Arg{template}\verb/{#1} }/
+%\end{quote}
+%    to process this key/value list. The outer test here, is only for
+%    speeding up the processing in case the list is empty.
```

Comment:
There isn't a lot in it - not sure the complexity is worth it!

Reply:
your and my understanding of the word "not much" clearly differs :-) Calling `\SetTemplateKeys` on an empty argument generates more than 500 lines of tracing compared to 11 for the test.

Reply:
I suspect the model of passing on a document level optional argument via `\SetTemplateKeys...{#1}` will be relatively common which means that calling this with an empty set might be more common than if you were directly setting keys "by hand". In which case, rather than suggesting an explicit test here we could consider optimising the empty argument case to reduce those 500 lines?

Reply:
yes, I agree with you both that it would be better to put it inside; I was documenting what we currently have and there this a) is the main/only real use case, b) it will appear in many/most document level commands/environments over time, and c) that argument will most of the time be empty

So yes, let's move it inside and simplify document and code.

Reply:
So should I wait for this change before reviewing?

Reply:
yes

Reply:
@car222222 would be ready now I think